### ISSUE TITLE
CupertinoActivityIndicator causes 200% CPU on iOS simulators

### ISSUE THREAD

**Issue by pedia** [flutter/flutter#13203]
## Steps to Reproduce

Create a widget below:
```dart
class Loading extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new Column(mainAxisAlignment: MainAxisAlignment.center, children: [
      new Center(child: const CupertinoActivityIndicator()),
      const Padding(padding: const EdgeInsets.only(top: 8.0)),
      new Center(child: new Text("Loading..."))
    ]);
  }
}
```

Then run on a simulator, "top -o cpu"

```
PID    COMMAND      %CPU
13241  backboardd   128.3
13661  Runner       99.7
```  

Then press 'q', the CPU drop to zero.

## Flutter Doctor

```
[✓] Flutter (on Mac OS X 10.13.1 17B48, locale en-HK, channel master)
    • Flutter at /Users/odd/wa/flutter/flutter
    • Framework revision bdbe53be01 (2 days ago), 2017-11-24 19:13:18 -0800
    • Engine revision 93b2179597
    • Tools Dart version 1.25.0-dev.11.0

[✓] Android toolchain - develop for Android devices (Android SDK 26.0.2)
    • Android SDK at /Users/odd/Library/Android/sdk
    • Unable to locate Android NDK.

    • Unable to locate compiler in Android NDK.

    • Platform android-26, build-tools 26.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] iOS toolchain - develop for iOS devices (Xcode 9.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.1, Build version 9B55
    • ios-deploy 1.9.2
    • CocoaPods version 1.3.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] Connected devices
    • iPhone X • 2B869FE0-C8D2-4755-9C69-D0F91CED8588 • ios • iOS 11.1 (simulator)
```

**Comment by eseidelGoogle** [comment]
@xster would know.

**Comment by pedia** [comment]
@eseidelGoogle @xster  A simple question should not open a new issue.
Why Cupertino Widgets are not applying material, eg: "Dark Visual Mode" not working.

**Comment by Hixie** [comment]
@pedia feel free to file new bugs for questions or feature requests like that.

**Comment by xster** [comment]
I can repro on iOS simulator with both Cupertino and the Material circular progress indicator. 
Didn't happen on physical device.

**Comment by xster** [comment]
Seeing same issue with animated gifs on simulator too

**Comment by eseidelGoogle** [comment]
@pedia did you mean to close this?

**Comment by xster** [comment]
This is an issue. Reopening.

**Comment by xster** [comment]
Did some testing. Nothing seemed like an immediate cause. We're issuing a fairly simple paint on a render object per frame. Likely something deeper in the engine.

**Comment by eseidelGoogle** [comment]
Could be related to https://github.com/flutter/flutter/issues/17889 ?

**Comment by liyuqian** [comment]
Flutter in iOS simulator uses Skia's software backend, which is very different from Skia's GPU backend---the one we used in actual iOS/Android devices---so it's probably not related to #17889 . 

From the tracing, it seems that both GPU thread and UI thread are quite fast:
<img width="1076" alt="screen shot 2018-10-05 at 3 41 36 pm" src="https://user-images.githubusercontent.com/22987568/46563264-4e1d9f00-c8b5-11e8-920c-bdfaace5fc13.png">

So this seems to be just caused by having an animation, not by how complicated or how costly the animation is. I verified with our [animation tutorial](https://raw.githubusercontent.com/flutter/website/master/src/_includes/code/animation/animate3/main.dart). I guess that the 200% CPU is just because that we have a GPU thread and a UI thread that are always running? @chinmaygarde 

Finally, I think that we don't have a severe performance issue because it's just running on iOS simulator. Again, iOS simulator and real iOS devices have very different performance characteristics so please don't test performance issues using iOS simulators.

**Comment by gunhansancar** [comment]
@liyuqian Personally I think your assumption about if it is simulator it is not important is really wrong. I'm having the same issue when on the widgets I've endless animations.

The problem with the simulator uses 200% cpu in processes (Runner, backboardd) causes a massive problem on my side. First it just eats out my battery. I cannot work without a charging cable at all because of this issue. Second my machine tends to be burning because of this huge . consumption. Third, electricity isn't free right :) Also on a real device, I noticed that massive battery consumption when you run an endless animation.

I believe this issue must be addressed but most probably the problem is in the core.

**Comment by liyuqian** [comment]
@gunhansancar : thanks for the feedback! I agree that it's important and we should fix it.

What I tried to convey is that a performance issue in the iOS simulator may not be necessarily a performance issue in real iOS devices. Usually developers care more about the real devices that their users are using (e.g., draining developer's battery has much less electricity impact than draining user's battery because there are probably millions of user devices per developer device).

Since you mentioned that you also noticed massive battery consumption on a real device, can you share us how you reproduce that problem and how you measure the battery consumption? If we can confirm that problem, it will be a much much more severe issue than just draining developer's laptop battery :smiley:

**Comment by gunhansancar** [comment]
@liyuqian You're definitely right about the scale of a single developer machine vs millions of devices of course :)

You must agree that even on the simulator this is not the expected usage. And I know there are priorities for the issues etc. etc. Just saying this should be on the radar and addressed early on IMHO.

So on a real device (iPhone 6) and with the release built I observe ,using Instruments Activity Monitor app on mac, average 70% cpu usage for the processes (60% Runner, 10% backboardd) when the animated widget on the screen. This also causes phone to get warm quickly and dry up the battery.

I will try to upload a simple project for you guys to test as well. We need to verify that if this is a general issue or a specific case.

**Comment by qwilbird** [comment]
+1 
This is a real issue for us too

**Comment by palancana** [comment]
Today I woke up with an idea in mind and turns out it works well, at least for now. It seems that both _CupertinoActivityIndicator_ and Material _LinearProgressIndicator_ makes _backboardd_ process of the iOS Simulator go crazy **just** when they are active. That means, when _CupertinoActivityIndicator_ `animating` property is true or _LinearProgressIndicator_ `value` property is null. If they are forced to have an static animation (`animating = false` or `value = 0.0`) the iOS Simulator behaves as normal.

As these are both activity load progress indicators and should not run for long periods, controlling that value means that we can work again with our computers without that heat and noise. I did the following on my project:
```dart
LinearProgressIndicator(value: loading ? null : 0.0)
```
Where `loading` is a `bool` variable that is true when i need it being animated and false when not. In my personal project it is controlled before and after an `http.get(url)`. I tried also the _CupertinoActivityIndicator_ and also works fine:
```dart
CupertinoActivityIndicator(animating: loading ? true : false)
```
I still think this should not be the normal behaviour, but maybe we can have this explained on the official widget documentation so we can spread the word and have people control the _backboardd_ process. 

What do you think of this?

**Comment by gunhansancar** [comment]
@palancana It is all about animations. If you stop the animation, it stops the excessive cpu usage. On my case, I'm doing background color animation continuously, so I can't stop it after certain time. But the simple color tweening and repaints shouldn't be caused that much trouble. 

I feel that this problem is about the rendering engine or in the core level somewhere but I can't verify that.

**Comment by dnfield** [comment]
From my local experiments, using the metal backend significantly improves this.

**Comment by liyuqian** [comment]
@dnfield : What's the CPU usage number in your experiments?

Maybe it's a good idea to turn on our metal backend by default first in all iOS simulators. It's less risky than launching on user devices, and it brings much more performance gains.

I think our metal backend is using the real GPU on the Mac instead of the CPU software backend. Can you confirm this, @chinmaygarde ?

**Comment by dnfield** [comment]
Sorry, I should clarify this - using th metal backend on a physical device helps this.  I forgot this bug was for simulator.

That said, if we could get the metal backend running on a simulator, that may help.

**Comment by edTheGuy00** [comment]
I discovered that the iOS simulators are just really bad with animations. I did the following:

1. Open an iOS simulator leave it idle - simulator runs smoothly
2. Open up system monitor, cpu usage is normal
3. Click and hold an App icon on the Simulator home screen to trigger the icons animation - watch as the CPU usage shoots up.

So the problem is 100% on the iOS simulators, however It looks like Flutter uses this type animation for everything including scrolling, and the longer the animation runs the worse it gets to the point where it just becomes unusable.

**Comment by kitenium** [comment]
This problem has made my dev machine fries like a chicken on the BBQ. Guys, seriously?!

**Comment by chinmaygarde** [comment]
Software rendering was selected as the default on iOS simulators because OpenGL was implemented via emulation on the simulators. Using emulated OpenGL there was slower than just using the Skia software renderer directly. On the other hand, [Metal uses host passthrough](https://developer.apple.com/documentation/metal/developing_metal_apps_that_run_in_simulator?language=objc) instead of emulation. We were working on support for Metal in Flutter and this just [landed on iOS devices](https://github.com/flutter/flutter/wiki/Metal-on-iOS-FAQ) and we are now working on the same [for the simulators as well](https://github.com/flutter/flutter/issues/52358). I attempted to run simple examples using the [experimental support for Metal on iOS simulators](https://github.com/flutter/flutter/issues/52358#issuecomment-611152143) and the energy usage was back down to device baselines. I can't think of anything more to do to improve the energy usage of the software renderer specifically and switching to emulated OpenGL would just be a straight up regression (performance wise as energy usage stays about the same). So this problem should go away after https://github.com/flutter/flutter/issues/52358 is addressed.

**Comment by markusaksli-nc** [comment]
Reproducible on the latest master `1.22.0-10.0.pre.450` (iOS 14 emulator).
<details>
 <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.22.0-10.0.pre.450, on Mac OS X 10.15.7 19H2 x86_64, locale en-GB)
    • Flutter version 1.22.0-10.0.pre.450 at /Users/markus/development/flutter_master
    • Framework revision 9d41ddcbac (13 minutes ago), 2020-10-01 08:01:58 +0200
    • Engine revision 612acf349e
    • Dart version 2.11.0 (build 2.11.0-180.0.dev)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/markus/Library/Android/sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: /Users/markus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/193.6626763/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Users/markus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/193.6626763/Android Studio.app/Contents
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

 
[✓] Connected device (5 available)            
    • Nevercode’s iPhone (mobile)         • b668e524315069f3db3661ac11ff1f66afafebdb • ios            • iOS 13.7
    • iPhone SE (2nd generation) (mobile) • B46F1F22-93B0-40C4-8DA7-1C9816E399F7     • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)
    • macOS (desktop)                     • macos                                    • darwin-x64     • Mac OS X 10.15.7 19H2 x86_64
    • Web Server (web)                    • web-server                               • web-javascript • Flutter Tools
    • Chrome (web)                        • chrome                                   • web-javascript • Google Chrome 85.0.4183.121

• No issues found!
```
</details>
<details>
 <summary>Minimal reproducible code sample</summary>

```dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Sample',
      home: Scaffold(
        appBar: AppBar(
          title: Text('Sample'),
        ),
        body: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            new Center(
              child: const CupertinoActivityIndicator(),
            ),
            const Padding(
              padding: const EdgeInsets.only(top: 8.0),
            ),
            new Center(
              child: new Text("Loading..."),
            ),
          ],
        ),
      ),
    );
  }
}
```
</details>

CPU usage drops from 60% to almost nothing when removing the animated widget and hot restarting.

@chinmaygarde What's the status on https://github.com/flutter/engine/pull/17881? There have been a few pings for you there.

**Comment by normano** [comment]
This occurs for Android Emulator using CircularProgressIndicator as well.

**Comment by byroncoetsee** [comment]
Simply adding :
```
Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(body:  CircularProgressIndicator()),
    );
  }
```
sends my machine's CPU through the roof. 

Running an iPhone 12 Pro Max simulator on iOS 14.5. 
Latest Flutter as of this date.

2017 issue hanging around in 2021... Classic.

**Comment by PhantomRay** [comment]
Still an issue for flutter 2.x on iOS simulator.

**Comment by AlexeyRK** [comment]
Have the same issue. Resolved it by changed the size of IOS simulator:
<img width="302" alt="Снимок экрана 2022-05-18 в 23 08 27" src="https://user-images.githubusercontent.com/104352769/169137548-bc11dbcd-34ac-4e98-986f-9e4f1114a721.png">

**Comment by vleeuwenmenno** [comment]
Is that really the fix? I am working on my app and I also get this on physical devices. My iPhone 14 Pro Max heats up like crazy when I have this code enabled:

![image](https://github.com/flutter/flutter/assets/46482143/a7d1e15c-49c0-49b8-b6c3-dfdd54de0b36)

Switching it to an icon fixes it for now. I don't know if this is related.

**Comment by darshankawar** [comment]
Using [code sample](https://github.com/flutter/flutter/issues/13203#issuecomment-701913890) and running it on iOS simulator with and without Impeller on latest stable version gives me below results:

1. With Impeller:


https://github.com/flutter/flutter/assets/67046386/5963dd56-6d3b-4866-bb72-57a510af8a94


2. Without Impeller:


https://github.com/flutter/flutter/assets/67046386/2fa83374-d28b-4e0f-9e15-b4ee990dbcfc

In short, with Impeller, the cpu usage is in mid 30%, while without it, the usage is under 20%.

@pedia or anyone in this thread If you could revisit this issue at your end using latest sdk version and confirm if you also see a drop in cpu usage or still get the same as originally reported ?

**Comment by pedia** [comment]
@darshankawar Great, Resolved.
